swarm/runtime/safe_paths.py

```
@@ -46,3 +48,32 @@ def validate_path_component(component: str, name: str = "path component") -> str
+    if not path:
+        raise ValueError(f"{name} cannot be empty")
+
+    p = Path(path)
+    if p.is_absolute():
+        raise ValueError(f"{name} must be relative, not absolute")
+
+    # Check for traversal using os.path.normpath logic
+    # We want to ensure the normalized path doesn't start with ..
+    normalized = os.path.normpath(path)
+    if normalized.startswith("..") or "/../" in normalized.replace("\\", "/"):
+        raise ValueError(f"{name} cannot traverse outside root")
```

Ruff (0.15.2)

[warning] Avoid specifying long messages outside the exception class

(TRY003)